File: flink-app/src/main/java/com/example/ctr/domain/service/CTRResultWindowProcessFunction.java
package com.example.ctr.domain.service;

import com.example.ctr.domain.model.CTRResult;
import com.example.ctr.domain.model.EventCount;
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.util.Collector;

public class CTRResultWindowProcessFunction
        extends ProcessWindowFunction<EventCount, CTRResult, String, TimeWindow> {

    @Override
    public void process(String productId, Context context, Iterable<EventCount> elements,
            Collector<CTRResult> out) {
        EventCount counts = elements.iterator().next();
        long impressions = counts.getImpressions();
        long clicks = counts.getClicks();
        long windowStart = context.window().getStart();
        long windowEnd = context.window().getEnd();

        CTRResult result = CTRResult.calculate(productId, impressions, clicks, windowStart, windowEnd);
        out.collect(result);
    }
}

File: flink-app/src/main/kotlin/com/example/ctr/domain/service/CTRResultWindowProcessFunction.kt
package com.example.ctr.domain.service

import com.example.ctr.domain.model.CTRResult
import com.example.ctr.domain.model.EventCount
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction.Context
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.util.Collector

class CTRResultWindowProcessFunction : ProcessWindowFunction<EventCount, CTRResult, String, TimeWindow>() {

    override fun process(
        key: String,
        context: Context,
        elements: Iterable<EventCount>,
        out: Collector<CTRResult>
    ) {
        val counts = elements.iterator().next()
        val result = CTRResult.calculate(
            key,
            counts.impressions,
            counts.clicks,
            context.window().start,
            context.window().end
        )
        out.collect(result)
    }
}
